Composer/packages/client/src/utils/qnaUtil.ts

```diff
@@ -78,6 +87,30 @@ export const reformQnAToContainerKB = (projectId: string, qnaFiles: QnAFile[]):
   return newQnAfiles;
 };
 
+export const copySourceQnAFilesOnOtherLocales = (
+  projectId: string,
+  qnaFiles: QnAFile[],
+  locales: string[]
+): QnAFile[] => {
+  const originalSourceQnAFiles = qnaFiles.filter((f) => f.id.endsWith('.source'));
+  if (originalSourceQnAFiles.length === 0) return qnaFiles;
+  const createdFiles: QnAFile[] = [];
+  for (let i = 0; i < originalSourceQnAFiles.length; i++) {
+    //source.{locale}.qna
+    const newSourceQnAFileIds = locales.map((l) => `${originalSourceQnAFiles[i].id}.${l}`);
+    for (let j = 0; j < newSourceQnAFileIds.length; j++) {
+      const createdFile = qnaUtil.parse(newSourceQnAFileIds[j], originalSourceQnAFiles[i].content);
+      if (!qnaFiles.find((f) => f.id === createdFile.id)) {
+        createFile(projectId, `${createdFile.id}.qna`, createdFile.content);
+        createdFiles.push(createdFile);
+      }
+    }
```

Comment:
My concern is the complexity. It's an O(n2) complexity which seems can be optimized.

Reply:
this part of code is used to migrate previous source.qna files to source.locales.qna files, which will be used only once. The most time consuming operation here is the qnaUtil.parse. We can optimize it by parsing each source file only once.